src/cr/cube/matrix.py

```diff
@@ -163,2555 +324,1553 @@ def table_margin_unpruned(self):
+        margin arrays.
         """
-        # --- an aggregate columns-dimension is not summable ---
-        if self._columns_dimension.dimension_type in (DT.MR, DT.CA):
-            return ()
+        return _BaseCubeResultMatrix.factory(
+            self._cube, self._dimensions, self._slice_idx
+        )
 
-        # --- inserted vectors are indexed using their *negative* idx, i.e. their
-        # --- "distance" from the end of the subtotals sequence. This insures their
-        # --- ordering tuple sorts before all base-columns with the same position while
-        # --- still providing an idx that works for indexed access (if required).
-        subtotals = self._columns_dimension.subtotals
-        neg_idxs = range(-len(subtotals), 0)  # ---like [-3, -2, -1]---
+    def _dimension_labels(self, dimension, order):
+        """1D str ndarray of name for each vector of `dimension`.
 
-        return tuple(
-            _InsertedColumn(
-                subtotal,
-                neg_idx,
-                self._unordered_matrix.table_margin,
-                self._base_rows,
-                self._base_columns,
-            )
-            for subtotal, neg_idx in zip(subtotals, neg_idxs)
+        These are suitable for use as row/column headings; labels for subtotals appear
+        in the sequence specified by `order`, such that labels are ordered to correspond
+        with their respective data vector.
+        """
+        return np.array(
+            [e.label for e in dimension.valid_elements]
+            + [s.label for s in dimension.subtotals]
+        )[order]
+
+    def _dimension_order(self, dimension, empty_idxs):
+        """1D np.int64 ndarray of signed int idx for each vector in `dimension`.
+
+        Negative values represent inserted-vector locations. Returned sequence reflects
+        insertion, hiding, pruning, and ordering transforms specified in `dimension`.
+        """
+        collation_method = dimension.collation_method
+
+        # --- this will become a significantly more sophisticated dispatch when
+        # --- sort-by-value is implemented in later commits. In particular, the
+        # --- signature for the sort-by-value collator differs from the payload and
+        # --- explicit-order collators. So we'll defer refactoring this until that
+        # --- collator is added. (sc)
+        display_order = (
+            ExplicitOrderCollator.display_order(dimension, empty_idxs)
+            if collation_method == CM.EXPLICIT_ORDER
+            else PayloadOrderCollator.display_order(dimension, empty_idxs)
         )
 
+        # --- Returning as np.array suits its intended purpose, which is to participate
+        # --- in an np._ix() call. It works fine as a sequence too for any alternate
+        # --- use. Specifying int type prevents failure when there are zero elements.
+        return np.array(display_order, dtype=int)
```

Comment:
what kind of problem you could have with an empty array? I mean if display_order is [] or () you shouldn'y have problems with float64 by default

Reply:
This array is used for indexing (in an `np.ix_()` call) so its values need to be integers (floats can't be used for indexing). The default `numpy` type is `float`, so if you don't specify a `dtype` and `display_order` is an empty sequence, numpy has no examples from which to infer type, and you end up with an array with `dtype == float`, which raises an error (or used to at least) in the later `np.ix_()` call.